File: app/src/main/java/org/cerion/tasklist/data/AppDatabase.java
package org.cerion.tasklist.data;

import android.content.Context;
import android.util.Log;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.TypeConverters;

@androidx.room.Database(entities = {TaskList.class, Task.class}, version = 1, exportSchema = false)
@TypeConverters(DateConverter.class)
public abstract class AppDatabase extends RoomDatabase {

    private static final String DATABASE_NAME = "tasks3.db";
    private static volatile  AppDatabase instance;
    private static final Object LOCK = new Object();

    public abstract TaskListDao taskListDao();
    public abstract TaskDao taskDao();

    public static AppDatabase getInstance(Context context) {
        if (instance == null) {
            synchronized (LOCK) {
                if (instance == null) {
                    instance = Room.databaseBuilder(context.getApplicationContext(), AppDatabase.class, DATABASE_NAME).allowMainThreadQueries().build();
                }
            }
        }

        return instance;
    }

    private static final DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US);
    public void log() {
        dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));

        List<TaskList> lists = taskListDao().getAll();
        Log.d(TAG, "---------------------- App Database ----------------------");

        for (TaskList list : lists) {
            Log.d(TAG, list.logString(dateFormat));
            logTasks(list.getId());
        }

        Log.d(TAG, "----------------------------------------------------------");
    }

    private final String TAG = AppDatabase.class.getSimpleName();
    private void logTasks(String listId) {
        List<Task> tasks = taskDao().getAllbyList(listId);
        for (Task task : tasks) {
            String listid = String.format("%1$-" + 43 + "s", task.listId);
            String id = String.format("%1$-" + 55 + "s", task.id);
            Log.d(TAG, "\t" + listid + " " + id + " " + task.title);
        }
    }

}

File: app/src/main/java/org/cerion/tasklist/data/AppDatabase.kt
package org.cerion.tasklist.data

import android.content.Context
import android.util.Log
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.TypeConverters
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import java.text.SimpleDateFormat
import java.util.*

@androidx.room.Database(entities = [TaskList::class, Task::class], version = 2, exportSchema = false)
@TypeConverters(DateConverter::class)
abstract class AppDatabase : RoomDatabase() {

    abstract fun taskListDao(): TaskListDao
    abstract fun taskDao(): TaskDao
    fun log() {
        dateFormat.timeZone = TimeZone.getTimeZone("UTC")

        val lists = taskListDao().getAll()
        Log.d(TAG, "---------------------- App Database ----------------------")

        for (list in lists) {
            Log.d(TAG, list.logString(dateFormat))
            logTasks(list.id)
        }

        Log.d(TAG, "----------------------------------------------------------")
    }

    private fun logTasks(listId: String) {
        val tasks = taskDao().getAllbyList(listId)
        for (task in tasks) {
            Log.d(TAG, "    " + task.logString(dateFormat))
        }
    }

    companion object {
        val TAG: String = AppDatabase::class.java.simpleName

        private const val DATABASE_NAME = "tasks.db"
        @Volatile
        private var instance: AppDatabase? = null
        private val LOCK = Any()

        val MIGRATION_1_2 = object : Migration(1, 2) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL("ALTER TABLE " + TaskList.TABLE_NAME + " ADD COLUMN deleted INTEGER DEFAULT 0 NOT NULL")
            }
        }

        fun getInstance(context: Context): AppDatabase? {
            if (instance == null) {
                synchronized(LOCK) {
                    if (instance == null) {
                        instance = Room.databaseBuilder(context.applicationContext, AppDatabase::class.java, DATABASE_NAME)
                                .allowMainThreadQueries()
                                .addMigrations(MIGRATION_1_2)
                                .build()
                    }
                }
            }

            return instance
        }

        fun generateTempId(): String {
            val rand = Random()
            val i = rand.nextInt() + (1L shl 31)
            return "temp_$i"
        }

        private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US)
    }

}
